pkg/cache/nodes.go

```diff
@@ -51,8 +70,35 @@ func equals(n1 *v1.Node, n2 *v1.Node) bool {
 	return common.Equals(n1Resource, n2Resource)
 }
 
-func (nc *NodeController) addNode(obj interface{}) {
-	log.Logger.Debug("node-controller: AddNode")
+func (nc *schedulerNodes) addExistingAllocation(pod *v1.Pod) error {
+	nc.lock.Lock()
+	defer nc.lock.Unlock()
+
+	if utils.IsAssignedPod(pod) {
+		if appId, err := utils.GetApplicationIdFromPod(pod); err == nil {
+			if schedulerNode, ok := nc.nodesMap[pod.Spec.NodeName]; ok {
+				schedulerNode.addExistingAllocation(&si.Allocation{
+					AllocationKey:    pod.Name,
+					AllocationTags:   nil,
+					Uuid:             string(pod.UID),
+					ResourcePerAlloc: common.GetPodResource(pod),
+					QueueName:        utils.GetQueueNameFromPod(pod),
+					NodeId:           pod.Spec.NodeName,
+					ApplicationId:    appId,
+					Partition:        common.DefaultPartition,
+				})
+			}
+		} else {
+			return err
+		}
+	}
+	return fmt.Errorf("cannot existing allocation because pod %s is not assigned to a node", pod.Name)
+}
+
+func (nc *schedulerNodes) addNode(obj interface{}) {
+	nc.lock.Lock()
```

Comment:
Can we move the locking down to the point were we need it to make the lock cycle as short as possible?
Just before we call AddNode()